### PR TITLE
Add lint warnings on SharedFlow operations that never complete

### LINKED ISSUE #2340: Flow toList could cause infinite suspension
In the rather heated discussion in a Telegram chat, there were complaints that calling toList() function on an infinite Flow (like the StateFlow for example) could cause an infinite suspension, therefore, producing a rather hard to catch bug.

For example, this code will block indefinitely:
```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
fun main() {
    var flow = MutableStateFlow(0)
    runBlocking{
      val list = flow.toList()
    }
}
```

There are several solutions to that problem:
* Introduce a default large limit to this operation and throw an error if the limit is exceeded.
* Throw error immediately for specific types of Flow which are known to be infinite.
* Update documentation to warn people that using terminal operations on infinite flows could lead to infinite suspensions.

All those solutions could be applied together.

The positive idea from the same discussion. It is possible to introduce something like

```kotlin
inline class LimitedFlow(val flow: Flow): Flow by flow
```
And define dangerous terminal operations on it. In this case operations like limit would provide type-safe wrappers. And the developer could explicitly declare the Flow as limited manually via something like `Flow::asLimited`.

### LINKED ISSUE #2368: SharedFlow didn't cancel or throw exception with callbackFlow
The exception didn't throw to parent scope or cancel itself while using SharedFlow with callbackFlow. I tried to cancel sharedflow in case of failure. Here is my code snippet  

```
fun adapt(call: Call<T>): SharedFlow<T> {
        val scope = CoroutineScope(dispatcher)
        return callbackFlow<T> {
            call.enqueue(object : Callback<T> {
                override fun onResponse(call: Call<T>, response: Response<T>) {
                    sendBlocking(response.body())
                    close()
                }

                override fun onFailure(call: Call<T>, t: Throwable) {
                    close(t)
                }
            })
            awaitClose()
        }.catch {
            /**
             *Here caught exception that threw in Failure case
             * Try to cancel Job so SharedFlow can terminate but It didn't work However catching exception here won't crash my app
             */
            scope.cancel()
        }.shareIn(scope, SharingStarted.Lazily, 1)
    }
```
I used it
```
runBlocking{

//this will get first output in case of success and then completed this coroutine
//But In case of Failure, this will always wait to get the first item and never completed

    val result = adapt(someCall).first()
}
```
Here is another way tried to accomplish the same thing

```
fun adapt(call: Call<T>): SharedFlow<T> {
        val scope = CoroutineScope(dispatcher)
        return callbackFlow<T> {
            call.enqueue(object : Callback<T> {
                override fun onResponse(call: Call<T>, response: Response<T>) {
                    sendBlocking(response.body())
                    close()
                }

                override fun onFailure(call: Call<T>, t: Throwable) {
                    close(t)
                }
            })
            awaitClose()
        }.shareIn(scope, SharingStarted.Lazily, 1).catch {
            /**
             *This catch block never invoked in case of failure
             * Try to cancel Job so sharedFlow can terminate its coroutine
             */
              currentCoroutineContext().cancel()
        }
    }
```
I used this as previous but this time
```
runBlocking{
//this will get first output in case of success and then completed this coroutine
//But In case of Failure App crashes
    val result = adapt(someCall).first()

}
```
sharedflow.collect|first()|take(1){}  stays always in resume state I want to cancel it in case of failure.

### PR DESCRIPTION
**Pull request by elizarov**
Fixes #2340
Fixes #2368

### CODE REVIEW

File: kotlinx-coroutines-core/common/src/flow/operators/Lint.kt [hunk 1]
```diff
@@ -79,4 +80,61 @@ public fun FlowCollector<*>.cancel(cause: CancellationException? = null): Unit =
...
+    message = "SharedFlow never completes, so this operator has no effect.",
+    level = DeprecationLevel.WARNING,
+    replaceWith = ReplaceWith("this")
+)
+@InlineOnly
+public inline fun <T> SharedFlow<T>.catch(noinline action: suspend FlowCollector<T>.(cause: Throwable) -> Unit): Flow<T> =
+    (this as Flow<T>).catch(action)
+
+@Deprecated(
+    message = "SharedFlow never completes, so this operator has no effect.",
```
Comment (slavonnet) [inline kotlinx-coroutines-core/common/src/flow/operators/Lint.kt:95]: maybe text ressource move to const? Any optimize  f for equal annotation dubles?

Comment (elizarov) [reply kotlinx-coroutines-core/common/src/flow/operators/Lint.kt:95]: What do you mean?

Comment (slavonnet) [reply kotlinx-coroutines-core/common/src/flow/operators/Lint.kt:95]: No way to reduce copy-paste / duplicate code / messages?

Comment (elizarov) [reply kotlinx-coroutines-core/common/src/flow/operators/Lint.kt:95]: There is no reason to. The code is easier to read this way, with messages right here.

File: kotlinx-coroutines-core/common/src/flow/operators/Lint.kt [hunk 2]
```diff
@@ -79,4 +80,61 @@ public fun FlowCollector<*>.cancel(cause: CancellationException? = null): Unit =
...
+    replaceWith = ReplaceWith("this")
+)
+@InlineOnly
+public inline fun <T> SharedFlow<T>.retryWhen(noinline predicate: suspend FlowCollector<T>.(cause: Throwable, attempt: Long) -> Boolean): Flow<T> =
+    (this as Flow<T>).retryWhen(predicate)
+
+@Deprecated(
+    message = "SharedFlow never completes, so this terminal operation never completes.",
+    level = DeprecationLevel.WARNING,
+    replaceWith = ReplaceWith("awaitCancellation()")
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/common/src/flow/operators/Lint.kt:118]: `awaitCancellation` seems to be unintended

Comment (elizarov) [reply kotlinx-coroutines-core/common/src/flow/operators/Lint.kt:118]: I really cannot figure out a better replacement. We can leave it without a replacement, though.

**Review by qwwdfsad — APPROVED** [review]
Nice precaution with `InlineOnly`.
Could drop a comment `// Error in 1.5.0` so we won't forget to do it

**Pull request merged**